CIP-0021 | Restrictions on transactions signed by hardware wallets

gabrielKerekes opened the pull request.

crptmppt commented:
For the future, please do not allocate a tentative CIP number in the title - hard to juggle the titles, informal references, and which are currently ongoing under that number

SebastienGllmt reviewed (approved):
Content makes sense. Is the plan to merge this CIP for now and then update it again after for Alonzo HW support for smart contracts?

gabrielKerekes commented:
> Is the plan to merge this CIP for now and then update it again after for Alonzo HW support for smart contracts?

Yes, it might also need to be updated when multisig/native script support is added to HW wallets, which will happen before Alonzo.

crptmppt commented on CIP-0021/CIP-0021.md:
```diff
@@ -0,0 +1,115 @@
+---
+CIP: 0021
```
21 (tentative)

crptmppt commented on CIP-0021/CIP-0021.md:
```diff
@@ -0,0 +1,115 @@
+---
+CIP: 0021
+Title: Restrictions on transactions signed by hardware wallets
```
the title is a bit misleading. Rather this CIP wants to establish some best practices for HW wallets.

gabrielKerekes replied:
Perhaps the intention is unclear and should be made clearer in the CIP. It does not comment on best practices, but lists limitations arising from hardware limitations of Trezor and especially Ledger.

The intended CIP audience are people integrating with HW wallets (developers of software wallets and similar tools), not HW wallet developers. In other words the CIP should define how you should build your transactions in order for them to be compatible (signable) with HW wallets - specifically Ledger and Trezor (not sure if there are other significant ones currently). In general, the rules should be a superset of restrictions for each HW wallet, so if you follow this CIP, you would be compatible with all HW wallets, and if you don't, you risk that some parties use HW wallets and will not be able to sign your transaction. This is especially important in context of transactions signed by multiple parties.

dcoutts replied:
I'd suggest using the term "interoperability" in the title, e.g. "Transaction guidelines for interoperability with hardware wallets".

gabrielKerekes commented:
We added “Optional empty lists and maps”, “Outputs” and “Auxiliary data” sections due to an issue reported to the Cardano LedgerJS project - https://github.com/vacuumlabs/ledgerjs-cardano-shelley/issues/125 - outputs with no multi-assets contained the empty multiasset map which caused a tx hash mismatch on Ledger which doesn’t include it. https://github.com/cardano-foundation/CIPs/pull/107/commits/81d23947c95008dd4b9e0dd6d7c46e56ba18f722

KtorZ commented on CIP-0021/CIP-0021.md:
```diff
@@ -0,0 +1,139 @@
+
+## Motivation
+
+Due to certain limitations of HW wallets, especially very small memory and a limited set of data types supported by Ledger, HW wallets are not able to process all valid transactions which are supported by Cardano nodes.
```
```suggestion
Due to certain limitations of hardware (abbrev. HW) wallets, especially very small memory and a limited set of data types supported by Ledger, HW wallets are not able to process all valid transactions which are supported by Cardano nodes.
```

gabrielKerekes replied:
Changed in https://github.com/cardano-foundation/CIPs/pull/107/commits/0d287f800481a3ca4cec23caf1a09b0b8f3ef0a4.

KtorZ commented on CIP-0021/CIP-0021.md:
```diff
@@ -0,0 +1,139 @@
+- The keys in every map must be sorted from lowest value to highest.
+- Indefinite-length items must be made into definite-length items.
+
+See the RFC for details.
```
:+1:

KtorZ commented on CIP-0021/CIP-0021.md:
```diff
@@ -0,0 +1,139 @@
+
+**Unsupported entries**
+
+The transaction body entry `6 : update` must not be included.
```
While I imagine this to be useful, I don't also see any reason to prohibit it at the hardware level? In particular, the genesis leaders are then prevented to use hardware devices for generating updates? This sounds quite limiting for something which we would definitely want to have extra security on?

gabrielKerekes replied:
AFAIK there hasn't been a request yet for HW wallets to support the update field and so HW wallets currently don't support it, thus it should be mentioned as a restriction.

KtorZ commented on CIP-0021/CIP-0021.md:
```diff
@@ -0,0 +1,139 @@
+
+**Integers**
+
+While the Cardano CDDL specification usually does not limit the size of integers, HW wallets only support `int64` for signed integers and `uint64` for unsigned integers. Any integer value must fit in the appropriate type.
```
Alas... I am still not sure whether unbounded integers on the blockchain was a good idea in the first place :no_mouth:

dcoutts replied:
We do limit the size in the CDDL, usually to signed or unsigned int 64. There are only a few places where we allow "big ints". See the `big_int` type definition in the Alonzo CDDL.

KtorZ replied:
> And note that those big ints are themselves bounded to 64 bytes.

That's good to know actually.

KtorZ commented on CIP-0021/CIP-0021.md:
```diff
@@ -0,0 +1,139 @@
+
+**Numbers of transaction elements**
+
+The numbers of inputs, outputs, certificates, withdrawals, tokens in an asset group, and other similar transaction elements must not exceed `UINT16_MAX`, i.e. 65535.
```
> and other similar transaction elements

Would it be possible to have an exhaustive list?

gabrielKerekes replied:
Added in https://github.com/cardano-foundation/CIPs/pull/107/commits/0d287f800481a3ca4cec23caf1a09b0b8f3ef0a4.

KtorZ commented on CIP-0021/CIP-0021.md:
````diff
@@ -0,0 +1,139 @@
+[ transaction_metadata: { * transaction_metadatum_label => transaction_metadatum }, auxiliary_scripts: [ * native_script ]]
+```
+
+The `auxiliary_scripts` must be an array of length 0.
````
Is that something which is in the making, or there's no plan to support auxiliary_scripts ever? That may be quite limiting for Plutus.

gabrielKerekes replied:
If `auxiliary_scripts` support is needed on HW wallets it should be added. Currently it's not supported and so it is a restriction on the transactions.

KtorZ commented on CIP-0021/CIP-0021.md:
```diff
@@ -0,0 +1,139 @@
+There are two limits on the number of witnesses:
+
+- an absolute limit of `UINT16_MAX`, i.e. 65535;
+- a relative limit dependent on the transaction body (essentially one witness per each input, each withdrawal and each certificate in a typical transaction).
```
This second condition may be overly restrictive in Alonzo where it'll be allowed to request additional signatures which do not correspond to any input. This would be however specified by a dedicated field on transactions `required_signers` (key 14 in transaction body).

gabrielKerekes replied:
This will change with the introduction of multi-sig to HW wallets which is currently being finalised. It will be possible to send `additional_witness_requests` to the HW wallets which will sign the transaction with a key derived from the path given in the request.

gabrielKerekes commented:
I've updated the CIP name and I've removed the sections regarding witnesses since this isn't true anymore. https://github.com/cardano-foundation/CIPs/pull/107/commits/74bb782218e68f713cba7735c4de5a522ee8216a

dcoutts reviewed (approved):
LGTM.

I'd again suggest correcting the section on integers. The CDDL does say where big ints vs fixed size ints are allowed. So the the Cardano CDDL *does* always limit the size of integers: either big ints are not included or where they are there is a size bound on them. If you see any place where that's not the case it's a bug that we should fix.

gabrielKerekes commented:
I've updated the section on integers in https://github.com/cardano-foundation/CIPs/pull/107/commits/f3d1b042404afac9cb2e444aac53b741a0429812.
